docs/design-archive/runtime-manager-design-doc.md

```diff
@@ -0,0 +1,116 @@
+        * [How to Register Plugins](#how-to-register-plugins)
+        * [Protocols between RuntimeManager and Plugins](#protocols-between-runtimemanager-and-plugins)
+        * [Examples for Runtime Plugins](#examples-for-runtime-plugins)
+    * [Installation](#installation)
+
+## Summary
+
+RuntimeManager acts as a proxy between kubelet and containerd(dockerd under dockershim scenario), which is designed to
+intercept CRI request, and apply some resource management policies, such as setting different cgroup parameters by pod
+priorities under hybrid scenario, applying new isolation policies for new linux kernel and latest cpu architecture etc.
+
+There are two components involved, RuntimeManager and RuntimePlugins.<br><br>
+![image](../images/runtime-manager-architecture.svg)
+
+## Goals
+
+- Enhance resource management for scenario like hybridization.
+- Provide interface for new isolation features which are not supported by CRI.
+
+## Components
+
+### RuntimeManager
+
+RuntimeManager is in charge of intercepting request during pod's lifecycle, such as RunPodSandbox, CreateContainer etc.,
+and then calling RuntimePlugins to do resource isolation policies before transferring request to backend containerd(
+dockerd)
+and after transferring response to kubelet. RuntimeManager provides an isolation-policy-execution framework which allows
+customized plugins registered to do specified isolation policies, these plugins are called RuntimePlugins.
+RuntimeManager itself does NOT do any isolation policies.
+
+### RuntimePlugins
+
+RuntimePlugins register events(RunPodSandbox etc.) to RuntimeManager and would receive notifications when events happen.
+RuntimePlugins should complete resource isolation policies basing on the notification message, and then response
+RuntimeManager, RuntimeManager would decide to transfer request to backend containerd or discard request according to
+plugins' response.
+
+If no RuntimePlugins registered, RuntimeManager would become a transparent proxy between kubelet and containerd.
+
+## Architecture
+
+![image](../images/runtime-manager-design.svg)
+
+There are mainly 4 components for RuntimeManager.
+
+### Cri Server
+
+as a proxy between kubelet and containerd, RuntimeManager acts as a CRI server for kubelet(http server under dockershim
+scenario). It should intercept all request from kubelet, and generate protocols for talking with plugins before and
+after talking with backend containerd(dockerd)
+
+### Plugins Manager
+
+PluginsManager is in charge of parsing registered plugin info from /etc/runtime/hookserver.d dynamically. examples for
+plugin register file:
+
+### Runtime Dispatcher
+
+RuntimeDispatcher is designed to manage communications with plugins.
+
+### Store
+
+As a proxy, RuntimeManager had better be designed as stateless, but sometimes it does NOT work. For StartContainer hook
+example, when StartContainer request comes, there exists only containerID in CRI StartContainerRequest, which is not
```

Comment:
`when StartContainer request comes, there exists only containerID in CRI StartContainerRequest` -> `there is only containerID in CRI StartContainerRequest`